ADC_DMA/Core/Src/main.c

```diff
@@ -0,0 +1,184 @@
+	uint16_t 		aux = 0;			//we only use aux to calculate v_ref
+
+	/*							  *
+	 * --- DMA INITIALIZATION --- *
+	 *							  */
+	RCC->AHBENR |= RCC_AHBENR_DMAEN;
+	DMA1_Channel1->CPAR	= 	(uint32_t) &(ADC1->DR);		//setting peripheral address
+	DMA1_Channel1->CCR 	|=	DMA_CCR_CIRC	|			//DMA in circular mode
+							DMA_CCR_MINC	|			//memory increment
+							DMA_CCR_PSIZE_0	|			//peripheral size set to 16 bits
+							DMA_CCR_MSIZE_0	;			//memory size set to 16 bits
+
+	/*										*
+	 * --- ADC INITIALIZATION PROCEDURE --- *
+	 *										*/
+	RCC->APB2ENR 	|= 	RCC_APB2ENR_ADC1EN;		//enables ADC clock
+	ADC1->CR 		|=	ADC_CR_ADVREGEN;		//enables the voltage regulator
+	ADC1->CFGR2 	|= 	ADC_CFGR2_CKMODE_0;		//sets ADC clock as PCLK/2
+	ADC->CCR 		|= 	ADC_CCR_LFMEN;			//low frequency mode enable
+	ADC1->SMPR 		|= 	ADC_SMPR_SMPR_1	|		//sets sample time for 12.5 ADC clock cycles
+						ADC_SMPR_SMPR_0	;		//because we need 10.5 ADC clock cycles at least
+	ADC1->CFGR1		|=	ADC_CFGR1_WAIT;			//enables wait mode to prevent overrun
+
+	ADC1->CR |= ADC_CR_ADCAL;					//starting the calibration
+	while(ADC1->CR & ADC_CR_ADCAL);				//we have to wait until ADCAL = 0 (Can be handled by interrupt)
+
+	ADC1->CFGR1	|=	ADC_CFGR1_DMACFG |			//DMA in circular mode
+					ADC_CFGR1_DMAEN	;			//enables DMA	- must be done AFTER calibrate the AD
+
+	ADC1->ISR|= ADC_ISR_ADRDY;					//clear the ADRDY bit by programming it to 1
+	ADC1->CR |= ADC_CR_ADEN;					//then we enable the ADC
+	while(!(ADC1->ISR & ADC_ISR_ADRDY));		//and wait for it to be ready (Can be handled by interrupt)
+
+	/*												*
+	 * --- ADC CHANNEL CONFIG FOR V_REF READING --- *
+	 *												*/
+	if(ADC1->CR & ADC_CR_ADSTART)					//we have to be sure there's no ongoing conversion
+	{
+		ADC1->CR |= ADC_CR_ADSTP;					//if so, we have to set ADSTP bit
+		while(ADC1->CR & ADC_CR_ADSTP);				//and wait for it to clear
+	}
+	ADC1->CR |= ADC_CR_ADDIS;						//disables the ADC
+	while(ADC1->CR & ADC_CR_ADEN);
+
+	ADC1->CHSELR = 	ADC_CHSELR_CHSEL17	|			//selecting the VREF channel
+					ADC_CHSELR_CHSEL18	;			//and Tsense channel
+	ADC->CCR 	|= 	ADC_CCR_VREFEN	| 				//enables internal reference voltage
+					ADC_CCR_TSEN	; 				//and temperature sensor
+	wait(TIME_10uSEC);								//we have to wait for the proper time for the Tsense to wake up
+
+	DMA1_Channel1->CNDTR = 2; 								//number of data to be transferred
+	DMA1_Channel1->CMAR = (uint32_t) &(ADC_measure.v_ref);	//set the first memory register address for DMA
+	DMA1_Channel1->CCR 	|= DMA_CCR_EN;						//enables the DMA
+
+	ADC1->CR |= ADC_CR_ADEN;						//then we enable the ADC
+	while(!(ADC1->ISR & ADC_ISR_ADRDY));			//and wait for it to be ready. (Can be handled by interrupt)
+
+	while(1)
+	{
+		ADC1->CR |= ADC_CR_ADSTART;						//starts the ADC
+
+		//we wait for the ADC and DMA to complete the process
+		while(!(ADC1->ISR & ADC_ISR_EOS) && !(DMA1->ISR & DMA_ISR_TCIF1)){}
+		wait(TIME_10uSEC);				//10 uS delay
```

Comment:
I don't like this delay, are you sure it is the only way to solve the problem with the memory address?

Reply:
The problem is I don't know exactly what's going on. It doesn't look like the problem is with the DMA itself.

I know that Vini can debug the code at home to see the error, I don't know if you have the PCB with you. I would tell you to remove this delay and compare the result after each run, with a breakpoint at nop line with running it step by step.

I'll read a little about it and think if I can use a different solution.

Reply:
Without the delay, just to show you what happens:

1st run, just a breakpoint at nop line:
![Screen Shot 2020-04-28 at 14 59 59](https://user-images.githubusercontent.com/61191550/80521159-efb98500-8960-11ea-9227-1c0dd991d1ce.png)

2nd run, without restarting the uC:
![Screen Shot 2020-04-28 at 15 00 39](https://user-images.githubusercontent.com/61191550/80521203-08299f80-8961-11ea-8474-038c06c0331c.png)

3rd run:
![Screen Shot 2020-04-28 at 15 01 15](https://user-images.githubusercontent.com/61191550/80521258-1d063300-8961-11ea-81de-41527d98ace9.png)

and it goes on...

now, if I just run the program step by step (using step into function on debugger), even without the delay, it works just fine

A single "nop" instead of the delay doesn't work either.

Reply:
Try clear ADC_measure.v_ref in line 82, before start the ADC again. Just a try, because I don't have the board here to test.

Reply:
Then I get v_ref right only in the first run. After that I always get 0.

And again, if I run the program step by step, or with the delay (and with v_ref = 0 before ADSTART), everything looks right. At least I get the right data

Reply:
Without the delay, if I put the breakpoint on line 88, 89 or 94, v_ref and temperature have the ADC data correctly (something around 1500 and 600) and the calculation's results are correct (I can see it with another breakpoint on nop line). But, if put a breakpoint on line 100, as well on nop line, I always get ADC data correctly, but after the calcs, I got that error

Reply:
Interesting, last try without studying about it, try make aux = 0, instead of v_ref = 0, at line 82.

Reply:
If I do just aux = 0, I get the same case that happens on the prints I uploaded yesterday. If I aux = 0; ADC_measure.v_ref = 0; ADC_measure.temperature = 0; on line 82 I get v_ref right only in the first run. After that, I always get 0 as before